### PR TITLE
Allow variables to be used when initing neighbor stateful props

### LINKED ISSUE #19735: Same issue of stateful mat prop with DG #11766
## Bug Description
We are trying to couple the DG and CG schemes for diffusion problems to enforce mass conservation. We started with adding a DG kernel to the diffusion CG kernel and ran the same input file of issue #11766 and got the following error:

```
Assertion `i < _size' failed

Access out of bounds in MooseArray (i: 0 size: 0)

at /Users/aalmetwally/projects/moose/framework/build/header_symlinks/MooseArray.h, line 276

Stack frames: 16

0: 0   libmesh_dbg.0.dylib                 0x000000010b6d8232 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 626

1: 1   libporous_flow-dbg.0.dylib          0x0000000100281b4d MooseArray<double>::operator[](unsigned int) const + 941

2: 2   libporous_flow-dbg.0.dylib          0x000000010037d586 PorousFlow1PhaseFullySaturated::buildQpPPSS() + 38

3: 3   libporous_flow-dbg.0.dylib          0x0000000100391858 PorousFlow1PhaseFullySaturated::initQpStatefulProperties() + 40

4: 4   libmoose-dbg.0.dylib                0x0000000103bfcdae MaterialBase::initStatefulProperties(unsigned int) + 94

5: 5   libporous_flow-dbg.0.dylib          0x0000000100390d8b PorousFlowMaterial::initStatefulProperties(unsigned int) + 107

6: 6   libmoose-dbg.0.dylib                0x0000000103d25487 MaterialPropertyStorage::initStatefulProps(MaterialData&, std::__1::vector<std::__1::shared_ptr<MaterialBase>, std::__1::allocator<std::__1::shared_ptr<MaterialBase> > > const&, unsigned int, libMesh::Elem const&, unsigned int) + 199

7: 7   libmoose-dbg.0.dylib                0x0000000104a01305 ComputeMaterialsObjectThread::onInternalSide(libMesh::Elem const*, unsigned int) + 1285

8: 8   libmoose-dbg.0.dylib                0x000000010408c4d4 ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool) + 1412

9: 9   libmoose-dbg.0.dylib                0x00000001042326f1 FEProblemBase::initialSetup() + 21329

10: 10  libmoose-dbg.0.dylib                0x0000000103b503ef Transient::init() + 2415

11: 11  libmoose-dbg.0.dylib                0x0000000105251516 MooseApp::executeExecutioner() + 1350

12: 12  libmoose-dbg.0.dylib                0x000000010524b44f MooseApp::run() + 8415

13: 13  porous_flow-dbg                     0x000000010000cfb5 main + 149

14: 14  libdyld.dylib                       0x00007fff20438f3d start + 1

15: 15  ???                                 0x0000000000000003 0x0 + 3

[0] /Users/aalmetwally/projects/moose/framework/build/header_symlinks/MooseArray.h, line 276, compiled Dec 10 2021 at 11:50:32

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0

[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1

:

system msg for write_line failure : Bad file descriptor

Process 14528 exited with status = 1 (0x00000001) 
```

## Steps to Reproduce

```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 2
  ny = 2
  xmin = 0
  xmax = 1
  ymin = 0
  ymax = 1
  elem_type = QUAD4
[]

[Variables]
  [./u]
    order = FIRST
    family = MONOMIAL
    [./InitialCondition]
      type = ConstantIC
      value = 1
    [../]
  [../]
[]

[Functions]
  [./forcing_fn]
    type = ParsedFunction
    value = 2*pow(e,-x-(y*y))*(1-2*y*y)
  [../]
  [./exact_fn]
    type = ParsedGradFunction
    value = pow(e,-x-(y*y))
    grad_x = -pow(e,-x-(y*y))
    grad_y = -2*y*pow(e,-x-(y*y))
  [../]
[]

[Kernels]
  [./diff]
    type = Diffusion
    variable = u
  [../]
  [./abs]          # u * v
    type = Reaction
    variable = u
  [../]
  [./forcing]
    type = BodyForce
    variable = u
    function = forcing_fn
  [../]
[]

[DGKernels]
  [./dg_diff]
    type = DGDiffusion
    variable = u
    epsilon = -1
    sigma = 6
  [../]
[]

[BCs]
  [./all]
    type = DGFunctionDiffusionDirichletBC
    variable = u
    boundary = '0 1 2 3'
    function = exact_fn
    epsilon = -1
    sigma = 6
  [../]
[]

[Materials]
  [./stateful]
    type = StatefulMaterial
    initial_diffusivity = 1
    boundary = 'left'
  [../]
  [./general]
    type = GenericConstantMaterial
    block = '0'
    prop_names = 'dummy'
    prop_values = '1'
  [../]
[]

[Executioner]
  type = Steady
  solve_type = 'PJFNK'
  nl_rel_tol = 1e-10
[]

[Postprocessors]
  active = 'h dofs l2_err'
  [./h]
    type = AverageElementSize
  [../]
  [./dofs]
    type = NumDOFs
  [../]
  [./l2_err]
    type = ElementL2Error
    variable = u
    function = exact_fn
  [../]
[]

[Outputs]
  file_base = out
  exodus = true
  csv = true
[]
```

## Impact
Cannot run DG model through the porousFlow module @lynnmunday @lindsayad

### PR DESCRIPTION
**Pull request by lindsayad**
Previously we did not reinit the variables so you would get a bad access
if you used a variable when initializing the neighbor version of
stateful material properties. I'm honestly a little surprised we never
hit this before

Closes #19735

### CODE REVIEW

File: framework/src/loops/ComputeMaterialsObjectThread.C [hunk 1]
```diff
@@ -171,7 +171,15 @@ ComputeMaterialsObjectThread::onInternalSide(const Elem * elem, unsigned int sid
 {
   if (_need_internal_side_material)
   {
-    _assembly[_tid]->reinit(elem, side);
+    const Elem * neighbor = elem->neighbor_ptr(side);
+    const dof_id_type elem_id = elem->id(), neighbor_id = neighbor->id();
```
Comment (fdkong) [inline framework/src/loops/ComputeMaterialsObjectThread.C:175]: Do we need to check if neighbor is empty?  Or this has been guarded by `_need_internal_side_material` ?

File: framework/src/loops/ComputeMaterialsObjectThread.C [hunk 2]
```diff
@@ -171,7 +171,15 @@ ComputeMaterialsObjectThread::onInternalSide(const Elem * elem, unsigned int sid
...
+    const bool should_continue =
```
Comment (fdkong) [inline framework/src/loops/ComputeMaterialsObjectThread.C:176]: It is a bit hard to understand this logic. Could you add comments ?

File: test/src/materials/VarCouplingMaterial.C [hunk 3]
```diff
@@ -21,12 +21,16 @@ VarCouplingMaterial::validParams()
...
   return params;
 }
 
 VarCouplingMaterial::VarCouplingMaterial(const InputParameters & parameters)
   : Material(parameters),
-    _var(coupledVectorTagValue("var", "tag")),
+    _var(getParam<bool>("use_tag") ? coupledVectorTagValue("var", "tag") : coupledValue("var")),
     _base(getParam<Real>("base")),
     _coef(getParam<Real>("coef")),
     _diffusion(declareProperty<Real>("diffusion")),
```
Comment (fdkong) [inline test/src/materials/VarCouplingMaterial.C:36]: This is not related to this PR. I was curious about why we call the material "diffusion"?

This should be a general material

**Review by fdkong — CHANGES_REQUESTED** [review]
Minor comments

File: framework/src/loops/ComputeMaterialsObjectThread.C (hunk 1 above)
Comment (lindsayad) [reply framework/src/loops/ComputeMaterialsObjectThread.C:175]: In `ThreadedElementLoopBase` we do not call `onInternalSide` if the neighbor is null

File: test/src/materials/VarCouplingMaterial.C (hunk 3 above)
Comment (lindsayad) [reply test/src/materials/VarCouplingMaterial.C:36]: I'm guessing it was probably because the test creator was coming up with a test where the material property was being consumed by a diffusion object

File: framework/src/loops/ComputeMaterialsObjectThread.C (hunk 2 above)
Comment (loganharbour) [reply framework/src/loops/ComputeMaterialsObjectThread.C:183]: Can we just add this logic as a utility somewhere if we use it multiple times? I think we use the same logic in DG? Maybe something on MooseMesh?

**Review by loganharbour — DISMISSED** [review]
Looks good to me otherwise. Up to you if you wanna cleanup duplication in skip logic (if there is duplication)

**Review by loganharbour — APPROVED** [review]
Beautiful code 👍🏼

**Comment by lindsayad** [comment]
@fdkong does this look fine for you? Please resolve conversations if they're resolved for you

**Review by fdkong — APPROVED** [review]
Nice improvement
